### PR TITLE
adds --filtering option to check.sh

### PR DESCRIPTION
**Pull request by astrale-sharp**
note that `check.sh itest --filtering clippy` wont parse clippy as a filter but as an arg

it shouldn't pose problems except maybe for fmt, I don't feel extremely confident improving the parsing in the sh language but if it's required i could rewrite it as a little rust script and pico-arg or something else.

this filters all tests.

### CODE REVIEW

**Comment by GodotRust** [comment]
API docs are being generated and will be shortly available at: https://godot-rust.github.io/docs/gdext/pr-308

File: itest/rust/src/runner.rs [hunk 1]
```diff
@@ -33,10 +33,17 @@ impl IntegrationTests {
...
     ) -> bool {
         println!("{}Run{} Godot integration tests...", FMT_CYAN_BOLD, FMT_END);
 
-        let (rust_tests, rust_file_count, focus_run) = super::collect_rust_tests();
+        let (rust_tests, rust_file_count, focus_run) = super::collect_rust_tests(
+            filters
+                .iter_shared()
+                .filter_map(|x| x.try_to().ok())
+                .collect::<Vec<_>>()
+                .as_slice(),
```
Comment (Bromeon) [inline itest/rust/src/runner.rs:45]: This could be its own variable.

**Review by Bromeon — COMMENTED** [review]
Thanks a lot!

I'd name it `--filter`, not `--filtering`. Could also come with a shortcut `-f`.

Now you run the filter logic twice, once in GDScript and once in Rust. Would it not be possible to have it in Rust only, and filter the `VariantArray` after passing it to Rust?

bors try

File: itest/godot/TestRunner.gd [hunk 2]
```diff
@@ -10,18 +10,13 @@ func _ready():
...
 			"--disallow-focus":
 				allow_focus = false
 			_:
-				unrecognized_args.push_back(arg)
-
-	if unrecognized_args:
-		push_error("Unrecognized arguments: ", unrecognized_args)
-		get_tree().quit(2)
-		return
+				filters.push_back(arg)
```
[the remark this answers is not in the document]
Comment (astrale-sharp) [reply itest/godot/TestRunner.gd:19]: >  Let's at least format it like [a,b,c]

I'm not sure what this means

maybe let's pass --filter or -f in gdscript too before it starts eating argument as filter?

File: check.sh [hunk 3]
```diff
@@ -159,6 +161,8 @@ function cmd_dok() {
 # between `itest` compilations and `check.sh` runs.
 extraCargoArgs=("--no-default-features")
 cmds=()
+itadd=false
```
[the remark this answers is not in the document]
Comment (astrale-sharp) [reply check.sh]: you're right about doing the filtering in rust, it's done there now!

would eatArgsAsFilter work?

File: check.sh [hunk 4]
```diff
@@ -34,11 +34,14 @@ Commands:
...
     -h, --help    print this help text
     --double      run check with double-precision
+    -f, --filter <args>     only run integration tests which contain any of the
+                            args (comma-separated). requires itest.
 
 Examples:
     check.sh fmt clippy
     check.sh
     check.sh --double clippy
+    chech.sh test itest -f variant ffi
```
Comment (Bromeon) [inline check.sh:44]: Could you make the implementation only accept comma-separated values, as I suggested in the help text?
```
    --filter <args>    only run integration tests which contain any of the
                       args (comma-separated). requires itest.
```


Otherwise you get the ambiguity of ordering, which can be unintuitive.
```suggestion
    chech.sh test itest -f variant,ffi
```
It should also simplify implementation that you would only need to consider the immediate next argument, and not any arguments following `--filter`.

File: itest/rust/src/runner.rs [hunk 5]
```diff
@@ -33,10 +33,22 @@ impl IntegrationTests {
...
+            .iter_shared()
+            .filter_map(|x| x.try_to().ok())
+            .collect::<Vec<_>>();
+        let gdscript_tests = gdscript_tests
+            .iter_shared()
+            .filter(|test| {
+                let test_name = get_property(test, "method_name");
+                passes_filter(filters.as_slice(), &test_name)
+            })
+            .collect::<Array<_>>();
```
Comment (Bromeon) [inline itest/rust/src/runner.rs:49]: Could also use type annotations for readability (2x):
```rs
let filters: Vec<ActualTypeHere> = filters.do_stuff().collect();
```

File: itest/godot/TestRunner.gd (hunk 2 above)
[the remark this answers is not in the document]
Comment (Bromeon) [reply itest/godot/TestRunner.gd]: I meant to enclose "filter arguments" in a special syntax. Do not make them dependent on argument ordering.

So instead of `a b c` or even `--filter a b c`, you should pass a single command-line argument `"[a,b,c]"`.

File: itest/rust/src/runner.rs [hunk 6]
```diff
@@ -33,10 +33,22 @@ impl IntegrationTests {
...
         allow_focus: bool,
         scene_tree: Gd<Node>,
+        filters: VariantArray,
     ) -> bool {
         println!("{}Run{} Godot integration tests...", FMT_CYAN_BOLD, FMT_END);
-
-        let (rust_tests, rust_file_count, focus_run) = super::collect_rust_tests();
+        let filters = filters
+            .iter_shared()
+            .filter_map(|x| x.try_to().ok())
```
Comment (Bromeon) [inline itest/rust/src/runner.rs:41]: This silently skips invalid elements, possibly hiding logic errors. Don't do that -- just assume all arguments have the correct type. Anything else is a bug that must be fixed, not ignored.

TLDR: use `.map(|v| v.to::<String>())`

**Review by Bromeon — COMMENTED** [review]
Thanks for the update! A few things left 🙂

File: itest/godot/TestRunner.gd (hunk 2 above)
[the remark this answers is not in the document]
Comment (astrale-sharp) [reply itest/godot/TestRunner.gd]: Okay, I finally understood!

**Comment by astrale-sharp** [comment]
This should be good to go!

**Comment by Bromeon** [comment]
Added a few things:
* Error message when `-f/--filter` is provided without having `itest` as a command
* Fix `--filtering` still being used

Thanks!
bors r+
